<a href="https://colab.research.google.com/github/YurleyG/Analitica-de-negocios/blob/main/reto1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Se procede con la carga de las librerias de trabajo

In [ ]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

# Se carga la base de datos

In [ ]:
nxl='/content/1. BD2_CarCrash_Int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)

ID      County            City  Weekday  Severity  ViolCat  \
0            1   SAN DIEGO       SAN DIEGO        7         1        8   
1            2    HUMBOLDT  UNINCORPORATED        4         1        8   
2            3     VENTURA          OXNARD        2         1       12   
3            4  STANISLAUS  UNINCORPORATED        4         1        1   
4            5   MENDOCINO  UNINCORPORATED        5         1        1   
...        ...         ...             ...      ...       ...      ...   
112655  112656      SHASTA         REDDING        1         0       11   
112656  112657      SHASTA         REDDING        6         1       11   
112657  112658      SHASTA         REDDING        5         0       11   
112658  112659      SHASTA         REDDING        3         0       11   
112659  112660      SHASTA         REDDING        5         1       11   

        ClearWeather  Month CrashType  Highway  Daylight  
0                  0      1         A        0         1  
1                  1      1         A        1         0  
2                  1      2         A        0         1  
3                  1      1         A        0         0  
4                  1      1         A        1         0  
...              ...    ...       ...      ...       ...  
112655             1      3         G        0         1  
112656             1      2         G        0         0  
112657             1      5         G        0         1  
112658             1      7         G        0         1  
112659             1      8         G        1         0  

[112660 rows x 11 columns]

1. ¿Cuál es el número de datos total de la base de datos?
Hay en total 112.660 datos

# Implementación del modelo

In [ ]:

#Separamos las variables por Entrada-Salida
XD=XDB[['Weekday','Severity','ViolCat', 'ClearWeather', 'Month','CrashType','Daylight']] #Corrected the typo from 'CrahType' to 'CrashType'
yd=XDB[['Highway']] #Aquí están las variables de referencia

# Crea un diccionario que mapea cada valor único de 'CrashType' a un número según el orden alfabético
crash_type_mapping = {crash_type: i for i, crash_type in enumerate(sorted(XDB['CrashType'].unique()))}

# Aplica el mapeo a la columna 'CrashType'
XD['CrashType'] = XD['CrashType'].map(crash_type_mapping) # Changed XDB to XD to apply mapping to the input features

# Fit the GaussianNB model
mnd=GaussianNB()
mnd.fit(XD,yd)



In [ ]:
from sklearn.naive_bayes import GaussianNB
mnb = GaussianNB()
mnb.fit(XD,yd)

#se obtiene la informacion
ct=mnb.class_count_
print("El total de los datos es:",len(XD))
print("los datos porcentualmente estan div:",ct/len(XD))
#Respuesta: Se encuentra divido en un 28% causado en autopista y un 71% no
#El total de los dato son 112660

El total de los datos es: 112660
los datos porcentualmente estan div: [0.7148855 0.2851145]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


2. ¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?
Los datos se encuentran divididos en un 71% de accidentes que no fueron en una autopista y 28% que si lo fueron

In [ ]:
mnd=GaussianNB()
mnd.fit(XD,yd)
ydp=mnd.predict (XD)


u=mnd.theta_
sigma=np.sqrt(mnd.var_)


LI=u-sigma
LS=u+sigma

print("la media de las variables por categoria es:")
print (u)
print("el limite inferior de la variación de las variables por categoria es:")
print(LI)
print("el limite superior de la variación de las variables por categoria es:")
print(LS)


cm=confusion_matrix(yd,ydp)


la media de las variables por categoria es:
[[3.90307801 0.06909696 6.79464607 0.89238754 6.67500217 2.80866413
  0.69460758]
 [3.92883161 0.07325426 4.45913888 0.86345382 6.68867719 2.49905047
  0.69487251]]
el limite inferior de la variación de las variables por categoria es:
[[ 1.96994846 -0.18452201  3.1928027   0.58249719  3.31451241  1.31524779
   0.23403413]
 [ 2.01569903 -0.18729916  1.98399159  0.52008618  3.35154117  1.29925827
   0.2344111 ]]
el limite superior de la variación de las variables por categoria es:
[[ 5.83620757  0.32271593 10.39648945  1.20227789 10.03549193  4.30208046
   1.15518103]
 [ 5.84196418  0.33380767  6.93428617  1.20682146 10.02581321  3.69884266
   1.15533393]]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


3.	Determinar los intervalos en los cual se encuentra la variación del día de la semana para cada una de las categorías de clasificación
[1.96  5.83  2.01  5.84]
4.	Determinar los intervalos en los cual se encuentra la variación de la ViolCat para cada una de las categorías de clasificación
[3.19  10.39  1.98  6.93]
5.	Determinar el intervalo en los cuales se encuentra la variación de la Categoría de accidentes para cada una de las categorías de clasificación
[1.31  4.30  1.29  3.69]


In [ ]:
ydp=mnb.predict(XD) #Hazme el pronostico de los positivos y negativos

from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

Ex=(VN+VP)/(VN+FP+FN+VP)
print("La exactitud es:",Ex)

TE=(FN+FP)/(VN+FP+FN+VP)
print("La tasa de error es:",TE)

Sen=VP/(VP+FN)
print("La sensibilidad es:",Sen)

Esp=VN/(VN+FP)
print("La especificidad es:",Esp)

pre=VP/(VP+FP)
print("El presición es:",pre)

pren=VN/(VN+FN)
print("El presición negativa es:",pren)

[[74939  5600]
 [28176  3945]]
La exactitud es: 0.7001952778270903
La tasa de error es: 0.29980472217290965
La sensibilidad es: 0.12281684879051089
La especificidad es: 0.9304684686921864
El presición es: 0.4133053954950236
El presición negativa es: 0.7267516850118799


In [ ]:
#Para evaluar un caso
XS=[[5,1,1,3,4,0]]
mnb.predict(XS)

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


array([1])

**Analísis**

El modelo tiene una exactitud del 70%, pero su baja sensibilidad (12%) indica que rara vez identifica correctamente los accidentes en autopistas. Aunque clasifica bien los accidentes fuera de autopistas (93% de especificidad), su precisión para detectar accidentes en autopistas es baja (41%). Esto sugiere un sesgo hacia la clasificación de "no autopista" lo que desmuestra que el modelo no es completamente confiable a la hora de determinar si dicho accidente se provocó en una autopista.